stoppable_workers.go

```diff
@@ -70,13 +104,16 @@ func (sw *StoppableWorkers) Add(worker func(context.Context)) {
 
 // Stop idempotently shuts down all the goroutines we started up.
 func (sw *StoppableWorkers) Stop() {
+	// Call `cancelFunc` with the write lock that competes with "readers" that can add workers. This
+	// guarantees `Add` worker calls that start a goroutine have incremented the `workers` WaitGroup
+	// prior to `Stop` calling `Wait`.
 	sw.mu.Lock()
-	defer sw.mu.Unlock()
 	if sw.ctx.Err() != nil {
 		return
```

Comment:
If we hit this return statement, we never unlock the mutex, and the program could deadlock. I think you're right that in this specific function, it's better to manually unlock the mutex and not use a `defer` statement, but usually I prefer using `defer` because it makes these things much easier to reason about.

Reply:
Oof yeah @dgottlieb I think we're missing an `Unlock` here. You're definitely right that `defer` normally handles this stuff, but we really wanted that `Wait` to be outside of a lock. Thanks for taking a look post-hoc @penguinland !

Reply:
Linking things up: I made https://github.com/viamrobotics/goutils/pull/361 to fix this.